providers/scaleway/provider.go

```diff
@@ -0,0 +1,255 @@
+			Tags:    pool.Tags,
+		}
+
+		resp, err := api.ListServers(&req, scw.WithContext(ctx))
+		if err != nil {
+			return nil, err
+		}
+
+		if resp.TotalCount > 0 {
+			// TODO(raskyld): add a warning if there are more than 1 found, it means there are orphan resources
+			return resp.Servers[0], nil
+		}
+	}
+
+	return nil, &InstanceDoesNotExists{
+		InstanceName: name,
+		Project:      *project,
+		Zones:        zones,
+	}
+}
+
+func (p *Provider) getAllInstances(ctx context.Context) ([]*instance.Server, error) {
+	pool := p.scwCfg.InstancePool[DefaultPool]
+	zones, err := pool.Locality.ResolveZones()
+	if err != nil {
+		return nil, err
+	}
+
+	api := instance.NewAPI(p.client)
+	instances := make([]*instance.Server, 0, 150)
+
+	for _, zone := range zones {
+		// TODO(raskyld): handle pagination for cases with more than 50 agents running per region
+		req := instance.ListServersRequest{
+			Zone:    zone,
+			Project: pool.ProjectID,
+			Tags:    pool.Tags,
+		}
+
+		resp, err := api.ListServers(&req, scw.WithContext(ctx))
+		if err != nil {
+			return nil, err
+		}
+
+		if resp.TotalCount > 0 {
+			instances = append(instances, resp.Servers...)
+		}
+	}
+
+	return instances, nil
+}
+
+func (p *Provider) createInstance(ctx context.Context, agent *woodpecker.Agent) (*instance.Server, error) {
+	pool := p.scwCfg.InstancePool[DefaultPool]
+	zones, err := pool.Locality.ResolveZones()
+	if err != nil {
+		return nil, err
+	}
+
+	// TODO(raskyld): Implement a well-balanced zone anti-affinity to spread instance
+	// 								evenly among zones for greater resilience.
+	random := rand.New(rand.NewSource(time.Now().Unix()))
+	zone := zones[random.Intn(len(zones))]
+
+	api := instance.NewAPI(p.client)
+
+	req := instance.CreateServerRequest{
+		Zone:              zone,
+		Name:              agent.Name,
+		DynamicIPRequired: scw.BoolPtr(true),
+		CommercialType:    pool.CommercialType,
+		Image:             pool.Image,
+		Volumes: map[string]*instance.VolumeServerTemplate{
+			"0": {
+				Boot:       scw.BoolPtr(true),
+				Size:       scw.SizePtr(pool.Storage),
+				VolumeType: instance.VolumeVolumeTypeBSSD,
+			},
+		},
+		EnableIPv6: pool.EnableIPv6,
+		Project:    pool.ProjectID,
+		Tags:       pool.Tags,
+	}
+
+	res, err := api.CreateServer(&req, scw.WithContext(ctx))
+	if err != nil {
+		return nil, err
+	}
+
+	return res.Server, nil
+}
+
+func (p *Provider) setCloudInit(ctx context.Context, agent *woodpecker.Agent, inst *instance.Server) error {
+	tpl, err := template.New("user-data").Parse(engine.CloudInitUserDataUbuntuDefault)
+	if err != nil {
+		return err
+	}
+
+	ud, err := engine.RenderUserDataTemplate(p.engineCfg, agent, tpl)
+	if err != nil {
+		return err
+	}
+
+	api := instance.NewAPI(p.client)
+
+	req := instance.SetServerUserDataRequest{
+		Zone:     inst.Zone,
+		ServerID: inst.ID,
+		Key:      "cloud-init",
+		Content:  bytes.NewBufferString(ud),
+	}
+
+	err = api.SetServerUserData(&req, scw.WithContext(ctx))
```

Comment:
This should be secure enough. :heavy_check_mark: 
I was wondering if this was secure, because on some clouds (Google), the user data (called metadata on Google) is accessible to all users on the instance, by just making a curl call.
https://www.scaleway.com/en/developers/api/instance/
> From a running Instance, by using the metadata API at http://169.254.42.42/user_data. To enhance security, we only allow user data viewing and editing as root. To know if the query is issued by the root user, we only accept queries made from a local port below 1024 (by default, non-root users can not bind ports below 1024). To specify the local port with cURL, use curl --local-port 1-1024 http://169.254.42.42/user_data